components/provider_cldr/src/transform/timezones/mod.rs

```diff
@@ -0,0 +1,366 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::cldr_langid::CldrLangID;
+use crate::error::Error;
+use crate::reader::{get_subdirectories, open_reader};
+use crate::CldrPaths;
+use icu_datetime::provider::{
+    key,
+    timezones::{
+        ExemplarCitiesV1, MetaZoneGenericNamesLongV1, MetaZoneGenericNamesShortV1,
+        MetaZoneSpecificNamesLongV1, MetaZoneSpecificNamesShortV1, TimeZoneFormatsV1,
+    },
+};
+use icu_provider::{
+    erased::ErasedDataProvider,
+    iter::{IterableDataProviderCore, KeyedDataProvider},
+    prelude::*,
+    serde::SerdeSeDataStruct,
+};
+use std::borrow::Cow;
+use std::convert::TryFrom;
+use std::marker::PhantomData;
+
+mod cldr_json;
+
+/// All keys that this module is able to produce.
+pub const ALL_KEYS: [ResourceKey; 6] = [
+    key::TIMEZONE_FORMATS_V1,
+    key::TIMEZONE_EXEMPLAR_CITIES_V1,
+    key::TIMEZONE_GENERIC_NAMES_LONG_V1,
+    key::TIMEZONE_GENERIC_NAMES_SHORT_V1,
+    key::TIMEZONE_SPECIFIC_NAMES_LONG_V1,
+    key::TIMEZONE_SPECIFIC_NAMES_SHORT_V1,
+];
+
+/// A data provider reading from CLDR JSON zones files.
+#[derive(PartialEq, Debug)]
+pub struct TimeZonesProvider<'d> {
+    data: Vec<(CldrLangID, cldr_json::LangTimeZones)>,
+    _phantom: PhantomData<&'d ()>, // placeholder for when we need the lifetime param
+}
+
+impl TryFrom<&dyn CldrPaths> for TimeZonesProvider<'_> {
+    type Error = Error;
+    fn try_from(cldr_paths: &dyn CldrPaths) -> Result<Self, Self::Error> {
+        let mut data = vec![];
+
+        let path = cldr_paths.cldr_dates()?.join("main");
+
+        let locale_dirs = get_subdirectories(&path)?;
+
+        for dir in locale_dirs {
+            let path = dir.join("timeZoneNames.json");
+
+            let mut resource: cldr_json::Resource =
+                serde_json::from_reader(open_reader(&path)?).map_err(|e| (e, path))?;
+            data.append(&mut resource.main.0);
+        }
```

Comment:
`suggestion (non-blocking)`: use map/collect on `locale_dirs` to produce `data`. I know it's CLDR so perf insensitive, so non blocking.

Reply:
Hmm, I don't think that the `?` operators would play nice with the closures in the iterator version here.

Reply:
It actually does (collect can collect into a `Result`), but not super pretty, so feel free to ignore: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=58fb5fe9eafce44cd07e2003bb1bfc51